library/Freckle/App/OpenTelemetry.hs

```diff
@@ -1,41 +1,139 @@
 module Freckle.App.OpenTelemetry
   ( HasTracer(..)
+  , Tracer
+
+  -- * Effects
   , MonadTracer(..)
   , inSpan
+
+  -- * Querying
+  , getCurrentTraceId
+  , getCurrentTraceIdHex
+
+  -- * Middleware
+  , addTraceIdThreadContext
+
+  -- * Setup
+  -- ** Built-ins
   , withTracerProviderXRay
+  , withTracerProviderXRayLogger
   , withTracerProviderDisabled
+
+  -- ** Combinatorially
+  , withTracerProvider
+  , setupAWSXRay
+  , setupLogger
+  , disableTracing
+
+  -- ** 'Tracer'
+  , makeTracer
+  , tracerOptions
   ) where
 
 import Freckle.App.Prelude
 
+import Blammo.Logging
+  ( LogLevel(..)
+  , Logger
+  , Message(..)
+  , logOtherNS
+  , runLoggerLoggingT
+  , withThreadContext
+  , (.=)
+  )
 import Control.Lens ((.~), (<>~))
+import Network.Wai (Middleware)
 import OpenTelemetry.AWSXRay
+import qualified OpenTelemetry.Context as Trace
+import qualified OpenTelemetry.Context.ThreadLocal as Trace
+import qualified OpenTelemetry.Logging.Core as Trace
 import OpenTelemetry.Trace
   ( HasTracer(..)
+  , Tracer
   , TracerProvider
+  , TracerProviderOptions
   , defaultSpanArguments
   , emptyTracerProviderOptions
+  , makeTracer
+  , tracerOptions
   )
+import qualified OpenTelemetry.Trace.Core as Trace
+  (SpanContext(..), getSpanContext)
+import OpenTelemetry.Trace.Id (Base(Base16), TraceId, traceIdBaseEncodedText)
 import OpenTelemetry.Trace.Monad (MonadTracer(..))
 import qualified OpenTelemetry.Trace.Monad as Trace
-import qualified OpenTelemetry.Trace.Setup as Trace
+import OpenTelemetry.Trace.Setup
 import OpenTelemetry.Trace.Setup.Lens
 
 inSpan :: (MonadUnliftIO m, MonadTracer m, HasCallStack) => Text -> m a -> m a
 inSpan = flip Trace.inSpan defaultSpanArguments
 
+addTraceIdThreadContext :: Middleware
+addTraceIdThreadContext app request respond = do
+  mTraceId <- getCurrentTraceIdHex
+
+  let
+    wrap = case mTraceId of
+      Nothing -> id
+      Just traceId -> withThreadContext ["trace_id" .= traceId]
+
+  wrap $ app request respond
```

Comment:
`traverse` by another name?

Reply:
I don't now! If you can rework to be, please do, I'd like to see it.